polars/polars-core/src/chunked_array/kernels/mod.rs

```diff
@@ -7,173 +7,56 @@ pub(crate) mod take_agg;
-where
-    S: PolarsPrimitiveType,
-    T: PolarsPrimitiveType,
-{
-    let data = arr.data();
-    let buf = data.buffers().to_vec();
-    let child_data = data.child_data().to_vec();
-    let (null_count, null_bit_buffer) = get_bitmap(arr);
-    let new_data = ArrayData::new(
-        T::DATA_TYPE,
-        arr.len(),
-        Some(null_count),
-        null_bit_buffer,
-        data.offset(),
-        buf,
-        child_data,
-    );
-    Arc::new(PrimitiveArray::<T>::from(new_data))
-}
-
-pub(crate) unsafe fn transmute_array_from_dtype<S>(
-    arr: &PrimitiveArray<S>,
-    dtype: ArrowDataType,
-) -> ArrayRef
-where
-    S: PolarsPrimitiveType,
-{
-    use ArrowDataType::*;
-    match dtype {
-        UInt64 => transmute_array::<_, UInt64Type>(arr),
-        UInt32 => transmute_array::<_, UInt32Type>(arr),
-        UInt16 => transmute_array::<_, UInt16Type>(arr),
-        UInt8 => transmute_array::<_, UInt8Type>(arr),
-        Int64 => transmute_array::<_, Int64Type>(arr),
-        Int32 => transmute_array::<_, Int32Type>(arr),
-        Int16 => transmute_array::<_, Int16Type>(arr),
-        Int8 => transmute_array::<_, Int8Type>(arr),
-        _ => todo!(),
-    }
-}
-
-/// This function will panic if the conversion overflows. Don't use it to cast to a smaller size.
-pub(crate) fn cast_numeric<S, T>(arr: &PrimitiveArray<S>) -> ArrayRef
-where
-    S: PolarsNumericType,
-    T: PolarsNumericType,
-    T::Native: num::NumCast,
-    S::Native: num::NumCast,
-{
-    let vals = arr.values();
-    let (_null_count, null_bit_buffer) = get_bitmap(arr);
-    let av = vals
-        .iter()
-        .map(|v| num::cast::cast::<S::Native, T::Native>(*v).unwrap())
-        .collect::<AlignedVec<T::Native>>();
-    Arc::new(av.into_primitive_array::<T>(null_bit_buffer))
-}
-
 pub(crate) fn is_nan<T>(arr: &PrimitiveArray<T>) -> ArrayRef
 where
     T: PolarsFloatType,
     T::Native: Float,
 {
-    let vals = arr.values();
-    let (_, null_bit_buffer) = get_bitmap(arr);
-    let mut builder = BooleanArrayBuilder::new_no_nulls(vals.len());
+    let validity = arr.validity();
 
-    vals.iter().for_each(|v| {
-        builder.append_value(v.is_nan());
-    });
-    let arr = match null_bit_buffer {
-        Some(buf) => builder.finish_with_null_buffer(buf),
-        None => builder.finish(),
-    };
-    Arc::new(arr)
+    let values = Bitmap::from_trusted_len_iter(arr.values().iter().map(|v| v.is_nan()));
+
+    Arc::new(BooleanArray::from_data(values, arr.validity().clone()))
 }
 
 pub(crate) fn is_not_nan<T>(arr: &PrimitiveArray<T>) -> ArrayRef
 where
     T: PolarsFloatType,
     T::Native: Float,
 {
-    let vals = arr.values();
-    let (_, null_bit_buffer) = get_bitmap(arr);
-    let mut builder = BooleanArrayBuilder::new_no_nulls(vals.len());
+    let validity = arr.validity();
+
+    let values = Bitmap::from_trusted_len_iter(arr.values().iter().map(|v| !v.is_nan()));
```

Comment:
Bitmap implements the `from_trusted_len_iter(Iter<bool>)`, which simplifies almost all operations that yield `bool`.